ufmt/core.py

```diff
@@ -2,118 +2,178 @@
+    result = usort(content, usort_config, path)
 
     if path.suffix == ".pyi":
-        mode.is_pyi = True
+        black_config.is_pyi = True
 
     try:
-        content = format_file_contents(content, fast=False, mode=mode)
-    except NothingChanged:
-        pass
-
-    return content
-
-
-def make_black_config(path: Path) -> Mode:
-    config_file = find_pyproject_toml((str(path),))
-    if not config_file:
-        return Mode()
-
-    config = parse_pyproject_toml(config_file)
+        content_str = result.output.decode(encoding)
+        content_str = format_file_contents(content_str, fast=False, mode=black_config)
+        content = content_str.encode(encoding)
+        return content
 
-    # manually patch options that do not have a 1-to-1 match in Mode arguments
-    config["target_versions"] = {
-        TargetVersion[val.upper()] for val in config.pop("target_version", [])
-    }
-    config["string_normalization"] = not config.pop("skip_string_normalization", False)
-
-    names = {
-        field.name
-        for field in Mode.__dataclass_fields__.values()  # type: ignore[attr-defined]
-    }
-    config = {name: value for name, value in config.items() if name in names}
-
-    return Mode(**config)
+    except NothingChanged:
+        return result.output
 
 
-def ufmt_file(path: Path, dry_run: bool = False, diff: bool = False) -> Result:
-    usort_config = UsortConfig.find(path)
-    black_config = make_black_config(path)
+def ufmt_string(
+    path: Path,
+    content: str,
+    usort_config: UsortConfig,
+    black_config: Optional[BlackConfig] = None,
+    encoding: Encoding = "utf-8",
+) -> str:
+    """
+    Format an arbitrary string value for the given path.
+
+    DEPRECATED: will be removed in µfmt version 3.0. Use :func:`ufmt_bytes` instead.
+    """
+    warn(
+        "ufmt_string will be removed in version 3.0; use ufmt_bytes instead",
+        DeprecationWarning,
+        stacklevel=2,
+    )
+
+    black_config = black_config or BlackConfig()
+
+    data = content.encode(encoding)
+    data = ufmt_bytes(
+        path,
+        data,
+        black_config=black_config,
+        usort_config=usort_config,
+        encoding=encoding,
+    )
+    return data.decode(encoding)
+
+
+def ufmt_file(
+    path: Path,
+    *,
+    dry_run: bool = False,
+    diff: bool = False,
+    black_config_factory: Optional[BlackConfigFactory] = None,
+    usort_config_factory: Optional[UsortConfigFactory] = None,
+) -> Result:
+    """
+    Format a single file on disk.
+
+    Passing ``dry_run = True`` will only format the file in memory, without writing
+    changes to disk. Passing ``diff = True`` will generate a unified diff of changes
+    on the :class:`Result` object.
+
+    Passing ``black_config_factory`` or ``usort_config_factory`` allows overriding the
+    default configuration for each respective tool. Must be pickleable functions that
+    take a :class:`Path <pathlib.Path>` and return a :class:`BlackConfig` or
```

Comment:
Worth mentioning in _paths that this gets absolute (or whatever word it actually is on Windows) paths, but _file passes it through as the user provides it?  Can we do the thing that trailrunner does for consistency?